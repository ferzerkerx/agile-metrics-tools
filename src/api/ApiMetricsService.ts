import { ElasticSearchService, TeamMetricsRequest } from '../domain/Types';
import { Logger } from '../domain/metrics/Logger';
import { MetricItem } from '../domain/metrics';
import { ExternalCollectorService } from '../domain/external';

export class ApiMetricsService {
  constructor(
    private readonly collectorService: ExternalCollectorService,
    private readonly metricsService: ElasticSearchService
  ) {}

  public async metricsForRequest(
    teamMetricsRequest: TeamMetricsRequest
  ): Promise<void> {
    try {
      Logger.info(`teamMetricsRequest: ${JSON.stringify(teamMetricsRequest)}`);
      const configurationDescriptors =
        ApiMetricsService.createConfigurationDescriptorsForRequest(
          teamMetricsRequest
        );
      Logger.info(
        `created configurations: ${JSON.stringify(configurationDescriptors)}`
      );
      for (const configurationDescriptor of configurationDescriptors) {
        const metricItems = await this.collectorService.fetch(
          configurationDescriptor.config
        );
        await this.pushMetrics(
          metricItems,
          configurationDescriptor.shouldUpdateEntries
        );
      }
      Logger.info('Done!');
    } catch (error) {
      Logger.error(error);
    }
  }

  public async pushMetrics(
    metricItems: MetricItem[],
    shouldReplaceEntry?: boolean
  ): Promise<void> {
    await this.metricsService.pushMetrics(metricItems, shouldReplaceEntry);
  }

  static createConfigurationDescriptorsForRequest(
    teamMetricsRequest: TeamMetricsRequest
  ): any[] {
    if (teamMetricsRequest.config) {
      return [{ ...teamMetricsRequest }];
    }

    throw Error(
      `Unable to create configuration for request ${JSON.stringify(
        teamMetricsRequest
      )}`
    );
  }
}
