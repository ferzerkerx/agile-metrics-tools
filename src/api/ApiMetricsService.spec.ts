import { ApiMetricsService } from './ApiMetricsService';
import {
  TeamMetricsRequestByService,
  TeamMetricsRequestByTeam,
} from '../Types';

describe('ApiMetricsService', () => {
  it('should construct default config', async () => {
    const teamMetricsRequest: TeamMetricsRequestByTeam = {
      shouldUpdateEntries: false,
      teamName: 'sprintteam',
    };
    const referenceDate = new Date('2019-02-10');
    const generatedConfigs = await ApiMetricsService.createConfigurationDescriptorsForRequest(
      teamMetricsRequest,
      referenceDate
    );
    expect(generatedConfigs).toMatchSnapshot();
  });

  it('should fail with invalid configuration', async () => {
    const teamMetricsRequest: TeamMetricsRequestByService = {
      serviceName: 'serviceName',
      shouldUpdateEntries: false,
      config: { fake: 'config' },
    };
    const referenceDate = new Date('2019-02-09');

    try {
      await ApiMetricsService.createConfigurationDescriptorsForRequest(
        teamMetricsRequest,
        referenceDate
      );
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it('should return existing unchanged config', async () => {
    const teamMetricsRequest: TeamMetricsRequestByService = {
      serviceName: 'serviceName',
      shouldUpdateEntries: false,
      config: { fake: 'config' },
    };
    const referenceDate = new Date('2019-02-12');
    const generatedConfigs = await ApiMetricsService.createConfigurationDescriptorsForRequest(
      teamMetricsRequest,
      referenceDate
    );
    expect(generatedConfigs).toMatchSnapshot();
  });
});
