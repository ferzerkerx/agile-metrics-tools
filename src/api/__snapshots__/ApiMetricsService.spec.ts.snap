// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiMetricsService should construct default config 1`] = `
Array [
  Object {
    "config": Array [
      Object {
        "estimateConfig": Object {
          "estimationValues": Array [
            1,
          ],
          "maxTime": 14,
        },
        "fields": Object {
          "storyPoints": "customfield_10005",
        },
        "since": "2019-02-10",
        "teamId": 123,
        "teamName": "sprintteam",
        "until": null,
        "workFlowMap": Object {
          "Closed": 8,
          "Code Review": 5,
          "Done": 9,
          "In Progress": 4,
          "On Hold": 2,
          "Open": 1,
          "Po Review": 6,
          "Ready For Dev": 3,
        },
        "workFlowType": "sprint",
      },
    ],
    "serviceName": "jira",
    "shouldUpdateEntries": false,
    "teamName": "sprintteam",
  },
  Object {
    "config": Array [],
    "serviceName": "github",
    "shouldUpdateEntries": false,
    "teamName": "sprintteam",
  },
  Object {
    "config": Array [],
    "serviceName": "sonar",
    "shouldUpdateEntries": false,
    "teamName": "sprintteam",
  },
  Object {
    "config": Array [],
    "serviceName": "jenkins",
    "shouldUpdateEntries": false,
    "teamName": "sprintteam",
  },
]
`;

exports[`ApiMetricsService should return existing unchanged config 1`] = `
Array [
  Object {
    "config": Object {
      "fake": "config",
    },
    "serviceName": "serviceName",
    "shouldUpdateEntries": false,
  },
]
`;
